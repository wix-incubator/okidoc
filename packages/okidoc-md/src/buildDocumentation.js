import buildDocumentationSource from './buildDocumentationSource';
import buildDocumentationMarkdown from './utils/buildDocumentationMarkdown';

function buildDocumentation({ title, entry, source, pattern, tag, visitor }) {
  const documentationSource = buildDocumentationSource({
    entry,
    source,
    pattern,
    tag,
    visitor,
  });

  return buildDocumentationMarkdown(documentationSource, {
    title,
    sourceInfo: entry || pattern,
  });
}

export default buildDocumentation;
