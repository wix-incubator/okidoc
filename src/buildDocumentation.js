import generate from '@babel/generator';
import documentation from 'documentation';

import buildDocumentationAst from './buildDocumentationAst';
import { buildMarkdown } from './output';

function buildDocumentation({ title, pattern, visitor }) {
  const { code: documentationSource } = generate(
    buildDocumentationAst(pattern, {
      visitor: visitor,
    }),
  );

  return documentation
    .build([{ source: documentationSource }], { shallow: true })
    .then(documentation.formats.json)
    .then(jsonString =>
      buildMarkdown({
        title: title,
        json: JSON.parse(jsonString),
      }),
    );
}

export default buildDocumentation;
