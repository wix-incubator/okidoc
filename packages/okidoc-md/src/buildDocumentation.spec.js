import buildDocumentation from './buildDocumentation';

describe('buildDocumentation', () => {
  it('should build documentation for class ', async () => {
    const sourceCode = `
      /**
      * @doc UI
      */
      class C1 {
        m1() {}
      }
      
      /**
      * @doc UI
      */
      export class C2 {
        m2() {}
      }
      
      /**
      * @doc UI
      */
      export default class C3 {
        m3() {}
      }
    `;

    expect(
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      }),
    ).toMatchSnapshot();
  });

  // waiting for @babel/generator fix deployed and published: https://github.com/babel/babel/pull/10258
  it.skip('should support class static interface', async function() {
    const sourceCode = `
      interface ClockConstructor {
        new (hour: number, minute: number): any;
        test(): string;
      }
      
      interface ClockInterface {
        tick(): any;
      }
      
      const Clock: ClockConstructor = class Clock implements ClockInterface {
        static test() { return 'string'; }
        constructor(h: number, m: number) {}
        tick() {
            console.log("beep beep");
        }
      }
    `;

    expect(
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      }),
    ).toMatchSnapshot();
  });

  it('should build documentation for functions ', async () => {
    const sourceCode = `
      /**
      * @doc UI
      */
      function functionDeclaration() {}
      
      /**
      * @doc UI
      */
      var varFunctionExpression = function () {}
      
      /**
      * @doc UI
      */
      let letFunctionExpression = function () {}

      /**
      * @doc UI
      */
      const constFunctionExpression = function () {}
      
      /**
      * @doc UI
      */
      const constArrowFunction = () => {}
      
      /**
      * @doc UI
      */
      export const exportConstArrowFunction = () => {}
      
      /**
      * @doc UI
      */
      export default exportDefaultConstArrowFunction = () => {}
    `;

    expect(
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      }),
    ).toMatchSnapshot();
  });

  it('should support jsx syntax', async function() {
    const sourceCode = `
      /**
       * @doc UI
       */
      class MyComponent extends React.Component {
        /**
         * Say Hello
         */
        sayHello() {
          
        }
        
        render() {
          return <div>hello</div>
        }
      }
    `;

    expect(
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      }),
    ).toMatchSnapshot();
  });

  it('should handle optional method in interface, issue #66', async () => {
    const sourceCode = `
      interface IRootContainerAPI {
        getWidth?(): number;
      }
    `;

    expect(
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      }),
    ).toMatchSnapshot();
  });

  it('should show readable error from `documentation.js`', async () => {
    // NOTE: https://github.com/niieani/typescript-vs-flowtype#bounded-polymorphism
    const sourceCode = `
      /**
      * @doc UI
      */
      function fooGood<T extends { x: number }>(obj: T): T {}
    `;

    expect.assertions(1);

    try {
      await buildDocumentation({
        source: sourceCode,
        tag: 'UI',
        title: 'Documentation',
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});
