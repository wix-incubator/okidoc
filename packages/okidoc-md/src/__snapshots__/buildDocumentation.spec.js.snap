// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildDocumentation should build documentation for class  1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Documentation

## C1

### m1()

## C2

### m2()

## C3

### m3()
"
`;

exports[`buildDocumentation should build documentation for functions  1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Documentation

## functionDeclaration()

## varFunctionExpression()

## letFunctionExpression()

## constFunctionExpression()

## constArrowFunction()

## exportConstArrowFunction()

## exportDefaultConstArrowFunction()
"
`;

exports[`buildDocumentation should handle optional method in interface, issue #66 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Documentation
"
`;

exports[`buildDocumentation should not fail when faced with \`new\` in type declaration, issue #59 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Documentation

## getClockClass()

> TODO: update markdown renderer #59

get clock class

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">RETURN VALUE</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>new</code>
        </td>
        <td>
            <div class=\\"function-type\\">function(<code>hour</code>: number, <code>minute</code>: number): any</div>
        </td>
      </tr>
      <tr>
        <td class=\\"param\\">
          <code>tick</code>
        </td>
        <td>
            <div class=\\"function-type\\">function(): string</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;

exports[`buildDocumentation should show readable error from \`documentation.js\` 1`] = `
[SyntaxError: "Documentation" documentation source - Unexpected token, expected , (2:19)
  1 | /** */
> 2 | function fooGood<T extends {
    |                   ^
  3 |   x: number;
  4 | }>(obj: T): T {}]
`;

exports[`buildDocumentation should support jsx syntax 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# Documentation

## MyComponent

### sayHello()

Say Hello

### render()
"
`;
